File: proposed-projects/round-trip-experiment/reverse/20260909T015142Z_f8ca89ff/20260910T053604Z_b86749e4/ported_implementation/src/rules-builder/symbol-ids.ts
export class SymbolIds {
  // we don't need to delete, just preserve relationships
  private map = new Map<string, number>();
  private pos = new Map<string, number>();
  private reverseMap = new Map<number, string>();

  public get size(): number {
    return this.map.size;
  }

  public keys(): string[] {
    return [...this.map.keys()];
  }

  public has(key: string): boolean {
    return this.map.has(key);
  }

  public get(key: string): number {
    const value = this.map.get(key);
    if (value === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return value;
  }

  public reverseGet(key: number): string {
    const value = this.reverseMap.get(key);
    if (value === undefined) {
      throw new Error(`SymbolIds does not contain value: ${key}`);
    }
    return value;
  }

  public getPos(key: string): number {
    const value = this.pos.get(key);
    if (value === undefined) {
      throw new Error(`SymbolIds does not contain key: ${key}`);
    }
    return value;
  }

  public set(key: string, value: number, pos: number): void {
    this.map.set(key, value);
    this.pos.set(key, pos);
    this.reverseMap.set(value, key);
  }

  public *[Symbol.iterator](): Generator<[string, number]> {
    yield* this.map.entries();
  }
}

File: packages/gbnf-experiment/data/20260909T035041Z_2d7fbef6/ported_implementation/gbnf/rules_builder/symbol_ids.py
from __future__ import annotations

from typing import Iterator

from ..utils.errors.gbnf_error import GBNFError


class SymbolIds:
    # we don't need to delete, just preserve relationships
    def __init__(self) -> None:
        self._map: dict[str, int] = {}
        self._pos: dict[str, int] = {}
        self._reverse_map: dict[int, str] = {}

    @property
    def size(self) -> int:
        return len(self._map)

    def __len__(self) -> int:
        return len(self._map)

    def keys(self) -> Iterator[str]:
        return iter(list(self._map.keys()))

    def has(self, key: str) -> bool:
        return key in self._map

    def __contains__(self, key: str) -> bool:
        return key in self._map

    def get(self, key: str) -> int:
        if key not in self._map:
            raise GBNFError(f"SymbolIds does not contain key: {key}")
        return self._map[key]

    def reverse_get(self, key: int) -> str:
        if key not in self._reverse_map:
            raise GBNFError(f"SymbolIds does not contain value: {key}")
        return self._reverse_map[key]

    def get_pos(self, key: str) -> int:
        if key not in self._pos:
            raise GBNFError(f"SymbolIds does not contain key: {key}")
        return self._pos[key]

    def set(self, key: str, value: int, pos: int) -> None:
        self._map[key] = value
        self._pos[key] = pos
        self._reverse_map[value] = key

    def __iter__(self) -> Iterator[tuple[str, int]]:
        yield from self._map.items()
